homeassistant/components/traccar_server/coordinator.py

```diff
@@ -123,12 +115,57 @@ async def _async_update_data(self) -> TraccarServerCoordinatorData:
                 "attributes": attr,
             }
 
-        if self.events:
-            self.hass.async_create_task(self.import_events(devices))
+        await self.subscribe()
 
         return data
 
-    async def import_events(self, devices: list[DeviceModel]) -> None:
+    async def handle_subscription_data(self, data: SubscriptionData) -> None:
+        """Handle subscription data."""
+        self.logger.debug("Received subscription data: %s", data)
+        self._should_log_subscription_error = True
+        update_devices = set()
+        for device in data.get("devices") or []:
+            device_id = device["id"]
+            if device_id not in self.data:
+                continue
+
+            if (
+                attr
+                := self._return_custom_attributes_if_not_filtered_by_accuracy_configuration(
+                    device, self.data[device_id]["position"]
+                )
+            ) is None:
+                continue
+
+            self.data[device_id]["device"] = device
+            self.data[device_id]["attributes"] = attr
+            update_devices.add(device_id)
+
+        for position in data.get("positions") or []:
+            device_id = position["deviceId"]
+            if device_id not in self.data:
+                continue
+
+            if (
+                attr
+                := self._return_custom_attributes_if_not_filtered_by_accuracy_configuration(
+                    self.data[device_id]["device"], position
+                )
+            ) is None:
+                continue
+
+            self.data[device_id]["position"] = position
+            self.data[device_id]["attributes"] = attr
+            self.data[device_id]["geofence"] = get_first_geofence(
+                self._geofences,
+                position["geofenceIds"] or [],
+            )
+            update_devices.add(device_id)
+
+        for device_id in update_devices:
+            dispatcher_send(self.hass, f"{DOMAIN}_{device_id}")
```

Comment:
We should use `async_dispatcher_send` in async context.

Reply:
It's possible to set coordinator data on a push event via:
https://github.com/home-assistant/core/blob/507384251429968ab2547aeed6786c33d6f0bac6/homeassistant/helpers/update_coordinator.py#L424

Reply:
That will update all listeners which is one of the things i did not want to do as it causes additional state “updates”

Reply:
Ok. Alternatively the coordinator supports adding a context when registering a listener which can be used to selectively update listeners. It's not without some extra code though. May not be worth it here.

Example:
https://github.com/home-assistant/core/blob/76784cbc957814efcf6761adc11cedcbb8e6dc4e/homeassistant/components/nibe_heatpump/coordinator.py#L32-L72

Reply:
Ill take a look at that after I am finished splitting all attributes/events into their own entities.